Add TypedArray support to jsi

wkozyra95 opened the pull request.

willholen committed:
Bump version to 0.5.0

Summary: This diff bumps the version to 0.5.0 in anticipation of the next release.

Reviewed By: tmikov

Differential Revision: D20950490

fbshipit-source-id: 6149ab78e49a454ed4c4cee933668fd1e77e896a

avp committed:
Clean up some includes in VM.

Summary: .

Reviewed By: tmikov

Differential Revision: D20877431

fbshipit-source-id: 72bc1c5c148982609b995f2d9e933457353fb7eb

avp committed:
Remove `using` statements in IR.h

Summary:
To avoid LLVM functions leaking into Hermes header files, remove
some `using` statements from a header file and just use `llvm::`
at the front of the casts (straightforward substitution).

Reviewed By: tmikov

Differential Revision: D20947252

fbshipit-source-id: 331e5a5dfc93bede598bebbb4dcd6f7d827768e5

willholen committed:
Let `hermesc` output a more helpful message by default

Summary:
Before this diff, running `hermesc` reads stdin and then fails
saying execution is unsupported. With this diff, it instead
says to choose output and gives an example.

Reviewed By: avp

Differential Revision: D20967165

fbshipit-source-id: c6349cbaffe1deaa43d7a5344e89f7e07e945c81

tmikov committed:
Enable the debugger by default in host builds

Summary:
Always enable the Hermes debugger in host builds to decrease the number
of supported configurations.

Reviewed By: willholen

Differential Revision: D20969039

fbshipit-source-id: 2ab5cb3aa5abf911ae60d20f0b8eb30bc04a3b3a

tmikov committed:
add config option to enable debugger

Summary:
Before this, the debugger was automatically enabled in dev and dbg
builds and disabled in opt builds. This diff adds a config flag
`hermes.debugger`, which, when set to True, will force the debugger to
be enabled.

Reviewed By: willholen

Differential Revision: D20959348

fbshipit-source-id: d3cf63d4276a7e5185cbac52bd7d3803588e357c

Commit:
Add Hermes CodeCoverage profiler API for checking enabled state

Summary: Add Hermes CodeCoverage profiler API for checking enabled state.

Reviewed By: dulinriley

Differential Revision: D20312546

fbshipit-source-id: 057a983dca2362bf31b8f1929f5ba3a658dd83c8

wkozyra95 committed:
Add TypedArray support to jsi

Commit:
Fix VMRuntimeTests for fbcode usage

Reviewed By: avp

Differential Revision: D20998411

fbshipit-source-id: 902332af39a96bcea6ba49fec708c33de1f7e212

Commit:
Fix TestHelpers createCodeBlock memory leak

Summary:
The `createCodeBlock` function from TestHelpers was used in a couple tests, but
didn't have a link between the `RuntimeModule` and the new `CodeBlock`.
This resulted in that `CodeBlock *` getting leaked.
The fix is simple: use `getCodeBlockMayAllocate` which properly initializes the
`RuntimeModule::functionMap_` entry, so that the destructor will free the code block memory.

Reviewed By: avp

Differential Revision: D20998412

fbshipit-source-id: f3375b16710f164721e0df947936a1680b16c26f

neildhar committed:
Delete copyLocale from RuntimeCommonStorage

Summary: These functions are not used anywhere, their inclusion was breaking my build.

Reviewed By: kodafb

Differential Revision: D21004785

fbshipit-source-id: 9a53fdf62f1ace3f6a5b74a9bd48d2579119b8af

neildhar committed:
Move definition of InputFilenames so it is visible in hermes.cpp

Reviewed By: avp

Differential Revision: D20978976

fbshipit-source-id: 191af2720e3e68ab69920d2f9629682fd20cbc1c

nikki93 commented:
Looked through this and the code looks good! One other case I was thinking about was: maybe in native code you want to access the underlying data of a `TypedArray` without any copies (so without doing a copy to a vector). This is useful if you are eg. immediately passing the data to some library that will then copy the data any way. For that scenario, you could get the `ArrayBuffer` then do `arrayBuffer->data()` to get a `uint8_t *` and then cast it to the element size and go from there (and keep the `uint8_t *` only while GC is still holding the `ArrayBuffer`). Maybe we could add a test case for it?

tmikov committed:
use "unsigned char" for bytecode string storage

Summary:
The bytecode string storage was typed as "char" and typically kept as
std::vector<char>, which was misleading since it contains both ASCII
and UTF16 values. It is a sequence of bytes, not chars. Parts of that
sequence can later be interpreted as one type or another.

Change it to use "unsigned char", which brings it in line with other
storage pools in the bytecode.

Perhaps more importantly, this makes it easier to use the string
deduping algorithms for binary data.

Reviewed By: avp

Differential Revision: D18056231

fbshipit-source-id: f97b7e5bbf94e82886bf92bea382ed7a9f592a0a

Commit:
Daily `arc lint --take CLANGFORMAT`

Reviewed By: zertosh

Differential Revision: D21037942

fbshipit-source-id: 2985609901c35b459fd6898bf2696cd97ad9a6ec

nikki93 commented:
Discussed re: my previous comment on Slack--`ArrayBuffer` already has a test for the data pointer + the case is covered through the `toVector` test (which internally does a copy), so looks good to me!

neildhar committed:
Update documentation to use new REPL tool

Reviewed By: avp

Differential Revision: D21000366

fbshipit-source-id: 82864f8b6ebcc8d2c821bf8550feaa4042ce730f

neildhar committed:
Add REPL features to Hermes CLI tool

Reviewed By: avp

Differential Revision: D20996738

fbshipit-source-id: 5e7652737daf4db2681c23752e08748f8e9d75e6

neildhar committed:
Require - for stdin input to hermes cli

Summary: Update tests and hermes.cpp so that '-' input is required to take stdin input.

Reviewed By: avp

Differential Revision: D20978967

fbshipit-source-id: 99eaf7341ba437287e6aa36e013686df51dda310

neildhar committed:
Update CircleCI npm job to use latest Ubuntu image (#215)

Summary:
The build is currently failing (possibly because 19.04 is EOL). This updates it to just use the latest Ubuntu.
Pull Request resolved: https://github.com/facebook/hermes/pull/215

Reviewed By: willholen

Differential Revision: D21026727

Pulled By: neildhar

fbshipit-source-id: 3b0bd7e286e8d2b21d3b7122e495c3fa31f9e41a

Commit:
hermes | Add explicit instantiations to ExternalStringPrimitive::create

Summary:
Without explicit instantiations clang-cl build on Windows cant link. It looks like they are required by C++ standard also?

> When an explicit instantiation names a class template specialization, it serves as an explicit instantiation of the same kind (declaration or definition) of each of its non-inherited **non-template members** that has not been previously explicitly specialized in the translation unit.

Reviewed By: tmikov

Differential Revision: D21047354

fbshipit-source-id: 6415cff13cf5b999c80028c17df36d81c0824a64

avp committed:
Add LLVM_LIKELY to branch in setNameLengthAndPrototype.

Reviewed By: mhorowitz

Differential Revision: D15340482

fbshipit-source-id: d1f2ea62b7ec06ca0b46a4ac6b77192b2fc8a2d5

poteto committed:
Capture IP while debugging

Summary: Fixes an issue where the current IP would not be set while the debugger is running. Also fixes the `throw-throw-eval.js` test which was previously expecting an inaccurate column number.

Reviewed By: avp

Differential Revision: D21044966

fbshipit-source-id: 9bde12f4ff7ea4f9241e13a45636c0664082ba5e

poteto committed:
Support .displayName property on Function in stack traces

Summary:
Adds support for `Function.displayName` in Hermes stack traces.

**This feature is non-standard and is not on a standards track. Do not use it in production. There may be large incompatibilities between implementations in other engines, and the behavior may change in the future.**

When defined with a non-empty string, the `displayName` property returns the display name of a function. All other types are ignored:

```
function a() {
  throw new Error('oops');
}

a.displayName = 'MyComponent';
a(); // Uncaught Error: oops
     //   at MyComponent

a.displayName = '';
a(); // Uncaught Error: oops
     //   at a

a.displayName = 123;
a(); // Uncaught Error: oops
     //   at a
```

`displayName` can also be set on function expressions:

```
let object = {
  someMethod: function() {
    throw new Error("oops");
  }
};

object.someMethod.displayName = 'anotherMethod';
object.someMethod(); // Uncaught Error: oops
                     //   at anotherMethod
```

Reviewed By: avp

Differential Revision: D21003447

fbshipit-source-id: b8a47556133d8f0616cd8b55f231d9a8a2cae4ce

poteto committed:
Support .displayName property on Function when debugging

Summary:
Adds support for `Function.displayName` in Hermes while debugging.

When defined with a non-empty string, the `displayName` property returns the display name of a function:

```
function first() {
  debugger;
  return;
}

function second() {
  first();
}

function third() {
  second();
}

first.displayName = "1st";
second.displayName = "2nd";
third.displayName = "3rd";

third();
```

Should show the following call stack in the debugger:

```
> 1st
  2nd
  3rd
  (anonymous)
```

Reviewed By: avp

Differential Revision: D21011071

fbshipit-source-id: 8bb0f4e54c859211d83926fa219fa8194f8d8ec2

Commit:
Add NoAllocScope around the whole snapshot

Summary:
It is an important property of a heap snapshot to require that no allocations happen as
a result of taking the snapshot.
Using `NoAllocScope` will help guarantee that (or at least give a debug assertion if an alloc
accidentally occurs).

Reviewed By: avp

Differential Revision: D21045788

fbshipit-source-id: 5277df596b4eb14b1f4443c9a885862570fa7fe5

Commit:
Prefer displayName over name for heap snapshots of functions

Summary:
In some contexts like debugging or exception messages,
`displayName` is prioritized over `name` on functions.
Add this prioritization to heap snapshots as well.

Reviewed By: avp

Differential Revision: D21045786

fbshipit-source-id: 1ed0ce19b74e48a25c4f4e514ff0f3427f05bf7b

Commit:
Simplify some conditions in heap snapshot name retrieval

Summary:
`dyn_vmcast` already makes sure that the HermesValue is a pointer before doing the
cast anyway, so it doesn't need the outer `isObject` or `isString` checks.

Reviewed By: avp

Differential Revision: D21045789

fbshipit-source-id: 5664baa24afd07c663bdad79ec6b2cef7ed7e8dc

avp committed:
Add marker flush in Debugger::getStackTrace.

Summary: This loop can allocate handles, so we should flush handles.

Reviewed By: tmikov

Differential Revision: D21068427

fbshipit-source-id: 0298adf106b94ba2118263d77b470d4fb0b32719

haozhun committed:
Rename and add trace related fields to RuntimeConfig

Summary:
Code that consumes these new fields are in an upcoming diff. The new fields are added, so that:
* synth trace will first be written to a temporary path, and then renamed at completion. This is necessary to avoid uploading (and deletion) in-progress synth trace files.
* The rewrite adds a finish callback to indicate completion of a synth trace file.

Reviewed By: dulinriley

Differential Revision: D20828842

fbshipit-source-id: c2bdf65fc7623c627d551800750b77ec55723021

haozhun committed:
Use TracingRuntime factory with rename and callback support

Summary: Utilize the new factory so that synth trace will first be written to a temporary path, and then renamed at completion.

Reviewed By: davedets

Differential Revision: D20749782

fbshipit-source-id: 790fbab2f20aa6aa23501d1c980f52a5a528b670

haozhun committed:
Extend TracingRuntime to support rename and callback

Summary:
* With the new API, synth trace will first be written to a temporary path, and then renamed at completion. This is necessary to avoid uploading (and deletion) in-progress synth trace files.
* It also adds a finish callback to indicate completion of a synth trace file.

Reviewed By: davedets, dulinriley

Differential Revision: D20828843

fbshipit-source-id: 0db9e42fa0ba71a710ae36b0eaa2a59c76ff065f

Commit:
Make chaos_until_crash.py script for use with Mozilla RR chaos mode

Summary:
Add a script to our debugging directory `lldb` that will run
Mozilla's RR ("Record and Replay") debugging tool multiple times on an
executable until a crash (non-zero exit code) occurs.

Takes `-r` or `--retries` to have a non-infinite maximum retry count.
Forwards stdout/stderr to the parent, so that all output can be monitored.

Once a failing trace is found, the tool exits.
The trace can be replayed with
```
rr replay /path/to/trace/directory
```

NOTE: In order to use `rr`, you must be on a Linux machine, and `perf list` must report the existence of Hardware Events.
VMs need to forward performance counters.
MacOS, Windows, and VMs without access to hardware profiling tools won't work.

Reviewed By: tmikov

Differential Revision: D21076062

fbshipit-source-id: 39a90ba3ce214395c814712e711a05eaf49e8688

tmikov committed:
add missing file Instrument.cpp

Summary:
Said file was missing from the build, but we didn't see it since IR
instrumentation is disabled by default in the CMake build.

Reviewed By: neildhar, willholen

Differential Revision: D21074735

fbshipit-source-id: 0b5b29a7ce0f6b1ba1a9f8130981cfcdb92445b8

ryantrem committed:
Fix jsi cmake include dirs (#207)

Summary:
I'm trying to use JSI for a React Native custom module. I saw these existing examples where the JSI API is used in the context of a CMakeLists.txt:
https://github.com/terrysahaidak/host-object-test/blob/master/libs/android-jsi/test-jsi/src/main/cpp/CMakeLists.txt
https://github.com/ericlewis/react-native-hostobject-demo/pull/4/files#diff-834320be1b4e4016bac27c05dcd17fb9
In both cases, they manually grab the include directories and jsi.cpp from node_modules/react-native, but I also saw that node_modules/react-native/ReactCommon/jsi/jsi already has a CMakeLists.txt that appears to be intended to provide a jsi static lib, so I tried to pull this into my own CMakeLists.txt like this:
```
add_subdirectory(${RN_DIR}/ReactCommon/jsi/jsi ${CMAKE_CURRENT_BINARY_DIR}/jsi)
...
target_link_libraries(MyLib jsi)
```
Unfortunately when doing this, the consuming project still doesn't see the correct include directories. The change I'm proposing here is to use `target_include_directories` and declare that `..` is a public (to consumers) include directory for the library named `jsi`. With this change, you can do what I showed above to consume the jsi lib by just pulling in the CMakeLists.txt file into your own CMakeLists.txt file.

Changelog: [General][Fixed] Fix jsi cmake include dirs

Pull Request resolved: https://github.com/facebook/hermes/pull/207

Reviewed By: willholen

Differential Revision: D21074270

Pulled By: tmikov

fbshipit-source-id: 7d9ec3255f57a16c0b2be489dffa4540727738a1

exced committed:
fix: define worker once (#218)

Summary:
https://github.com/facebook/hermes/issues/217

Fix website playground which is hanging forever.
worker var was reassigned in component and thus the play button was stuck on preheating Hermes.

rickhanlonii you mentioned you wanted it to work with SSR ? I might be missing details here 😄
Pull Request resolved: https://github.com/facebook/hermes/pull/218

Reviewed By: rickhanlonii

Differential Revision: D21082564

Pulled By: tmikov

fbshipit-source-id: 7a4298cb13f9776fcc828f6869cdcd0d4ffb1644

Commit:
DeclareGlobalVar to not overwrite existing properties (#220)

Summary:
Ha ha, this is a doozy.

Prior to this change, DeclareGlobalVariable would unconditionally store
undefined as a property on global. With this change, only do so if there is
not already such an own-property.

The reason for the fix is that eval() can otherwise overwrite existing
global variables. For example:

var q = true;
eval('var q;');

it is wrong for 'q' to result in undefined (and in practice, v8 and jsc do not).
Pull Request resolved: https://github.com/facebook/hermes/pull/220

Test Plan: Added a new test

Reviewed By: kodafb

Differential Revision: D21083640

Pulled By: tmikov

fbshipit-source-id: 73185ab8531ebea8f42dc637e103fea47507b2b4

Commit:
Patch fix for GTEST_HAS_EXCEPTIONS for Windows clang builds

Summary:
gtest 1.8.0 has a bug in detecting if exceptions are enabled when building
on Windows with clang.
Specifically, it relies on `_HAS_EXCEPTIONS` being defined by the STL, but libc++
doesn't set this, even though `_MSC_VER` is defined.

This was fixed in https://github.com/google/googletest/commit/3498a1ac52deb83f30b8170c78bfba9dc6227198,
which already exists in gtest 1.10.0: this diff pulls in that patch to 1.8.0.

Added to the README.md file to explain this patch.

Reviewed By: tmikov

Differential Revision: D21096393

fbshipit-source-id: 710df355349fc7026106e33edf3bea5651064638

neildhar committed:
Update tests and delete REPL code

Reviewed By: willholen

Differential Revision: D21000367

fbshipit-source-id: 39942050234fb72f82160eb8ae09ac0a18452a0b

rickhanlonii committed:
Add website README.md with deploy instructions (#223)

Summary:
## Overview

This PR adds a README to the `website/` directory that briefly explains the website and provides deploy instructions.
Pull Request resolved: https://github.com/facebook/hermes/pull/223

Reviewed By: dulinriley

Differential Revision: D21142472

Pulled By: tmikov

fbshipit-source-id: 420fa3a705a17447bd102ab5acce2946a055f6c8

tmikov committed:
return a copy of the bytecode instead of a view

Summary:
The compile function was incorrectly returning a view of the compiled
code instead of a copy.

Reviewed By: cpojer

Differential Revision: D21083581

fbshipit-source-id: ad3388fc101b1ef84c7fcbf7d4b9c33e45332ece

tmikov committed:
export JS API to validate a bytecode module

Summary:
The API performs basic sanity check of the supplied bytecode buffer:
- alignment
- buffer size
- magic and version

Reviewed By: cpojer

Differential Revision: D21083580

fbshipit-source-id: 4c2771e8f82632cf6fc5435a8ea6dfaa3b83ec18

Commit:
Improve sanitizer support in CMake builds (#224)

Summary:
This makes the `--enable-asan` option to `configure.py` work again, and adds an analogous `--enable-ubsan`. It also disables misaligned reads in miniz when run under UBSan, which was the one test failure. With this change, the hermes tests pass with both ASan and UBSan (gcc-flavored).
Pull Request resolved: https://github.com/facebook/hermes/pull/224

Reviewed By: neildhar

Differential Revision: D21142577

Pulled By: tmikov

fbshipit-source-id: 163d0ed613f3600da0f808b755afec46c70c22b2

tmikov committed:
dynamic optimization callback

Summary:
Until now BytecodeProviderFromSrc only invoked the IR optimization
passes if `compileFlags.optimize == true`, and
HERMESVM_ENABLE_OPTIMIZATION_AT_RUNTIME is defined to true. The intent
was to avoid linking the optimization passes in the VM if we don't need
them at runtime.

There are two problems with this:
- The API behaves inconsistently. A caller might request optimizations,
but they will not be performed depending on a build flag which the
caller may not even be aware of.
- If the build flag is not set, it is impossible to perform
optimizations at all. To get reliable results one is forced to copy the
implementation of BytecodeProviderFromSrc.

To solve this, we simply add a new static factory function accepting a
callback to perform optimizations. This defers the decision of whether
to link the optimization passes to the caller. Then can pass `nullptr`
(or an empty function), or they can pass `hermes::runAllOptimzations`,
or something in between.

For now preserve the existing unreliable API (only now it calls the new
one), but the hope is to be able to remove it soon.

Reviewed By: kodafb

Differential Revision: D20811492

fbshipit-source-id: 72f125705fea2204fe04fa3982746b7901005ef4

avp committed:
Place ifndef HERMESVM_LEAN on LazySource.

Summary:
The definition of the struct should have the same conditionals as its
usage.

Reviewed By: dulinriley

Differential Revision: D20967750

fbshipit-source-id: beac8bdfc6dfa06f412d6633e3a6e93ce6c6ffe7

Commit:
Miscellaneous small QoL improvements (#226)

Summary:
This is a grab-bag of some miscellaneous small changes.
The PR squashes these three separate commits:

## Improve commenting in NativeConstructor
This adds some comments in NativeConstructor, clarifying the role of certain
parameters.

## Pass context to CreatorFunction

It may be that a NativeConstructor's CreatorFunction needs the user-supplied
context pointer associated with the Callable. Pass that into the function.

Enhance a test to verify the address.

## HermesValueTraits::arrow() to guard against rvalue-references

HermesValueTraits<HermesValue> knows how to convert a
`const HermesValue &value` into a `const HermesValue *` by taking its
address.

Prior to this fix, if  you were to pass a temporary (rvalue-reference)
HermesValue in, it would be converted to a temporary const reference
and its address would be taken, returning a dangling pointer.
I forget the exact circumstances where this can happen but it
was very painful to debug.

Protect against this by adding a `HermesValue&&` overload which returns
`void` and is unimplemented.

Pull Request resolved: https://github.com/facebook/hermes/pull/226

Reviewed By: avp

Differential Revision: D21192781

Pulled By: tmikov

fbshipit-source-id: f97bcfec1118230678b9fbc7988bb90c06376df5

mhorowitz committed:
Allow use of std::tuple<> with decorators directly

Summary:
Previously, a derived class, WithTuple, was used.  This ran
into bugs in MSVC (see https://github.com/microsoft/STL/issues/121).
Instead, use specialization to get the same result using std::tuple
directly.  This avoids the bug, and is a cleaner API.

Changelog: [Internal]

Reviewed By: dulinriley

Differential Revision: D21233677

fbshipit-source-id: 1d75991847164e525b4ba70f65a90627e5f8cd56

Commit:
Introduce DecoratedObject and rebuild HostObject on top of it (#225)

Summary:
This PR conceptually factors out the "holds a native pointer" aspect of
HostObject, allowing it to be used independently from the "interposes properties."
The new intermediate class is called DecoratedObject.

This fills an integration gap.  Say you wish to attach a native thing to a JS object,
and manipulate the native bits from prototype functions. This is not possible
with HostObject, because the prototype is not searched. DecoratedObject is
the missing intermediate layer: it manages the native bits (memory management,
etc) without affecting the object model.

This PR squashes the following commits:

## Switch HostObject to storing unique_ptr instead of shared_ptr

HostObject currently stores a shared_ptr, but in practice the refcount is
always 1; jsi does not share objects. Switch to storing unique_ptr instead.
This both reduces allocation size and prepares for introducing
DecoratedObject.

## Introduce DecoratedObject

Hermes currently has no easy way to associate a finalizable pointer with an
JSObject. This commit introduces a new JSObject subclass, called
DecoratedObject, to fill this gap.

A DecoratedObject holds a unique_ptr to a Decoration, which is intended to
be subclassed. There are ways of getting and setting the Decoration, and
also for the Decoration to report its memory usage.

The Decoration is purely for "integration" uses. JSObject does not care about
the decoration.

The next commit proposes rebuilding HostObject atop DecoratedObject, which
reduces some duplication.

## HostObject to subclass DecoratedObject

This "rebuilds" HostObject on top of DecoratedObject. HostObjectProxy now
subclasses DecoratedObject::Decoration. The advantage is to reduce code
duplication: HostObject can take advantage of the memory management
of DecoratedObject.

The third commit rebuilds HostObject on top of DecoratedObject; now HostObject only does the property-interposing parts.
Pull Request resolved: https://github.com/facebook/hermes/pull/225

Reviewed By: tmikov

Differential Revision: D21211233

Pulled By: mhorowitz

fbshipit-source-id: 2e3e1100f00ed765df85be19231744445c328f1c

neildhar committed:
Use non-interactive option with apt-get (#229)

Summary:
Builds are currently failing because the tzdata package requires user
input while being installed via apt-get. Setting the environment
variable `DEBIAN_FRONTEND` to `noninteractive` forces apt-get to
bypass the requirement.
Pull Request resolved: https://github.com/facebook/hermes/pull/229

Reviewed By: tmikov

Differential Revision: D21242342

Pulled By: neildhar

fbshipit-source-id: 8f393ff785ec1f67e7ca9628dbb0cd85f2fff095

neildhar committed:
Use apt-get in non-interactive mode for all jobs (#230)

Summary:
This change avoids future build failures due to user input
requirements when installing packages via apt-get.
Pull Request resolved: https://github.com/facebook/hermes/pull/230

Reviewed By: tmikov

Differential Revision: D21244023

Pulled By: neildhar

fbshipit-source-id: ce694c6e9b7de42a9459afe9fe2479386bbeb51c

Commit:
Treat Strings as pointer types.

Summary: This is the squash of already approved diffs D19897285 and D21186651; they need to be squashed to make sure there isn't a point where MobileLab tests didn't work.

Reviewed By: dulinriley

Differential Revision: D21194012

fbshipit-source-id: 980ebb3381da5a8d0442dd0cb950231d71c1ad7d

Commit:
Add a "build properties" section to the synth trace.

Summary: There are several preprocessor flags, and other build attributes, that can affect the behavior of Hermes in the original execution.  This diff starts to record the values of those in the trace, so we can know that how we have to build Hermes used in replay.

Reviewed By: dulinriley

Differential Revision: D20818921

fbshipit-source-id: d513207e9953144af1a43e76b32a37f120b4adc8

Commit:
Add ability to get unique IDs for PropNameIds (= SymbolIDs).

Summary: On D19897285, Riley made a suggestion that was in fact a significant simplification: assigning ObjectID unique IDs to PropNameIDs, using the same mechanism we use for objects.  This diff creates the infrastructure for this.

Reviewed By: dulinriley

Differential Revision: D21009213

fbshipit-source-id: 947f3716718799421db73d4a9ca0139c3d395674

Commit:
Introduce mode to simulate Android in Linux desktop synth replay.

Summary:
For convenient playback, we want to be able to replay a trace taken on 64-bit android on 64-bit desktop Linux.  This diff introduces a config parameter and preprocessor variable governing that, and uses that preprocessor variable to fix the one place I found where the size of a class was different in 64-bit Android vs. Linux builds.  Apparently, the platform we use on Linux has a small-string optimization, so std::string is 32 bytes, whereas it is only 8 bytes on Android.  This makes the size of ExternalStringPrimitive differ between the two platforms.

To make them the same for replays on Linux, we add a config parameter indicating we are trying to simulate Android.  ExternalStringPrimitive, when compiled with this mode, uses a pointer to the string, instead of having the string inline.  This makes the sizes the same.  (As a comment notes, if this made the size too small, then we would add padding.)

Reviewed By: tmikov

Differential Revision: D20821130

fbshipit-source-id: a125e5d116e07da1b16163b60cfa15860c37b265

Commit:
Explicitly flush trace in synth when we're tracing.

Summary: Haozhun's recent diffs changed the behavior of TracingHermesRuntime, so that its destructor no longer flushes the trace -- flushing must be done explicitly.  When we run synth with -trace, this wasn't being done.  This diff fixes that.

Reviewed By: haozhun

Differential Revision: D21193854

fbshipit-source-id: 73b859d6ba4cd334d61b0b3b06c68c9d60e40c8d

Commit:
Add bool to BeginExecJS record indicating whether input was bytecode.

Summary:
The BeginExecJSRRecord type in synth trace records the hash of the JS source, but it doesn't record whether the original execution was from source, or from the bytecode compiled from that source.  This diff adds a boolean property to the record to indicate that.

We will probably use this in the script that collects the source or bytecode from the build artifacts, so it can find the .hbc or .js files, appropriately.

Reviewed By: dulinriley

Differential Revision: D20818603

fbshipit-source-id: 5d05cdbd5786061c8d6b673c1619880c9ce34d39

Commit:
Keep a record of GC executions.

Summary:
To determine whether we have precisely reproduced the sequence of GCs (and allocations) between the original execution and a trace, we need to keep a record of those GCs (and allocations) in the trace.  This diff does that.  Note that fine-grained allocations are only tracked prior to the first YG GC for now -- these would make the trace very large.

Currently, this is only used when we compile with a new hermes.api_trace_debug config parameter.  In the future, we may record the GC sequence always, and extend it to have the set of Object IDs reachable from external roots, so that we can only use the external roots observed during the original execution.

Reviewed By: dulinriley

Differential Revision: D20448892

fbshipit-source-id: 3ff945a07a281d760aea1afa0900d1f692264719

Commit:
Small simplification in some operator== methods.

Summary: During the course of working on SynthTrace, I noticed a couple of small simplifications that I couldn't resist.  This diff separates them out from a larger diff.

Reviewed By: dulinriley

Differential Revision: D20828785

fbshipit-source-id: 95fe4acc8bfe1f1dab9e0e2893866d6057eea946

Commit:
Test demonstrating that replay does extra string allocs.

Summary: I knew theoretically that the way we recorded strings could lead to more string allocations during replay than occurred during original execution.  This diff adds a test that demonstrates this problem.  Adds a new version of TraceInterpreter::execFromMemoryBuffer that takes an existing runtime, to make the test easier to write.

Reviewed By: dulinriley

Differential Revision: D20608474

fbshipit-source-id: 1ac544e45d7364d51d0f12244ec85e0f6b9f21c0

Commit:
Use the TracingRuntime when creating a PropNameID in marker function.

Summary: A tiny change: when we create the function for adding marker records, use the TracingRuntime to create the PropNameID for the global name of this function.

Reviewed By: dulinriley

Differential Revision: D20821067

fbshipit-source-id: 0606e58a6a56394eb5c0b783b30d27538a0f04e6

Commit:
Update configure.py to work with both upstream and fastcomp backends

Summary:
A recent update of the upstream Emscripten SDK no longer
allows the linker arg `-s BINARYEN_TRAP_MODE=clamp`.
This flag was only getting used because `EMSCRIPT_FASTCOMP` defaulted
to `ON`.

Change the `configure.py` script to always specify whether fastcomp is used
to avoid this issue.

Reviewed By: kodafb

Differential Revision: D21216950

fbshipit-source-id: b76fc8c6af3345f7a50994fbc44bf237ae235ae5

neildhar committed:
Add ability to dump sampling profiler output to stream

Summary:
Expose the ability to dump sampling profiler output to a given
stream. This makes it possible to directly extract sampling
information via the API without needing to write it to a file

Reviewed By: haozhun

Differential Revision: D20954418

fbshipit-source-id: 02890671894500674b7da07f82792c59f86372e5

willholen committed:
Allow specifying an input source map

Summary:
This diff lets you specify an input source map in the emhermesc JS interface.
All locations related to the module, whether in debugging or exceptions, will
report the source mapped location.

Reviewed By: tmikov

Differential Revision: D21100142

fbshipit-source-id: 0cba30e43af6e90c3feb421004f47687f33a03c6

neildhar committed:
Remove special build directory suffixes in configure.py

Summary:
Right now, `configure.py` generates different directory suffixes
depending on the build mode. This can lead to some slightly
surprising behaviour when people build for ASAN/UBSAN and can also
make automated scripts that use `configure.py` clunkier. This change
removes those special suffixes when a build directory is specified.

Reviewed By: dulinriley

Differential Revision: D21268445

fbshipit-source-id: 90f5497f3930dee35dbea7e8f7ad64d6ac244fca

neildhar committed:
Remove common.py

Summary:
Since there is now only a single python script to configure the
build, there is no longer a need to pull common functions out into
a separate script. This change consolidates all the required
functions into configure.py and removes common.py.

Reviewed By: dulinriley

Differential Revision: D21280915

fbshipit-source-id: 348f22eeafa9953e99dede76a16c162d50eda6f0

Commit:
Improve preprocessor flag for using fbsystrace

Summary:
Add a specific flag for avoiding defining fbsystrace on certain platforms.
Emscripten is one of those platforms (if `HERMES_FACEBOOK_BUILD` is defined).

Reviewed By: tmikov

Differential Revision: D21311166

fbshipit-source-id: e6315290af643702244ecf438a007bc8ef276872

Commit:
Fix build warning for lossy casts in HermesValue

Summary:
A new clang warning showed up when building for WASM:
```
implicit conversion from 'int64_t' to 'double' changes value
```
fix this by making the conversion explicit, and add a cast.

Reviewed By: avp

Differential Revision: D21311164

fbshipit-source-id: b317d1d3109d21c2a4212fefdd17bf58aa904e10

Commit:
Use specified int sizes in format strings

Summary:
Using `PRIdPTR` and `size_t` didn't work on all platforms or compilers.
Instead, use specified sizes like `uint64_t` so that `PRIu64` can be used
reliably.

This was only done in the GC.

Reviewed By: avp

Differential Revision: D21311167

fbshipit-source-id: e35c6dcfab8abb720ea0cff7141a29c8c23ee877

Commit:
Add Windows with clang Sandcastle job

Summary:
Windows allows building with clang on Sandcastle, and we should make sure
Hermes builds and passes tests when that mode is enabled.

#utd-hermes

Reviewed By: tmikov

Differential Revision: D21066599

fbshipit-source-id: baf0b2517e5192b15bfdf651b4d673a8a5cc8202

Commit:
Remove unnecessary sigaction in repl for WASM

Summary:
When running the hermes repl in WASM via node.js, there is no need for
special signal handling of `^C`, as node handles that for us.
Additionally, calling `sigaction` in WASM emits a warning message:
```
Calling stub instead of sigaction()
```
which is annoying to see on every line of the REPL.

Reviewed By: tmikov

Differential Revision: D21315419

fbshipit-source-id: 5c9238976e6780f1b78f04a36590c31a7de5bf96

mhorowitz committed:
Prevent internal header dependencies from sneaking into external JSI headers

Summary:
hermes/hermes.h should not depend on VM internals.  Remove
GCExecTrace.h, and replace it with a forward declaration.

To prevent this from happening in the future, refactor the
hermes_api_library macro to remove include:include from exported_deps.

Reviewed By: dulinriley

Differential Revision: D21308540

fbshipit-source-id: 4685c247518162e9108f259d808a8b2c4f0b2044

Commit:
Remove the non-Chrome memory profiler

Summary:
Hermes had an older memory profiler implementation attempt that targeted a non-Chrome
memory profiler format. Now that there exists a Chrome-based format for a memory profiler
that doesn't re-use any of these mechanisms, it's better to delete the unused code.

Reviewed By: avp

Differential Revision: D21261119

fbshipit-source-id: cbe709854e0567e5435c7ddce04031f078d7fa53

Huxpro committed:
Rename IdentifierTranslation to IdentifierHashes

Summary:
The field "identifierTranslations" is used to store the hashes
of identifiers. It was renamed from "hashes" to "translations" in
D14457157 and now we basically need to rename it back, including
related fields, variables and comments.

Reviewed By: tmikov

Differential Revision: D21336820

fbshipit-source-id: fea83d649d08093daca4c9ded41def5f01481078

Commit:
Implement scaffold for HadesGC

Summary:
Add some stubs so that a new HadesGC can be built and linked against the VM.
At this point, all tests immediately fail. The point of this diff is only to setup the build system
to allow it to be built.

Reviewed By: davedets

Differential Revision: D19729046

fbshipit-source-id: 116d3d2498287f020e687c679f997a4769fc8517

willholen committed:
Make column match fuzzy

Summary:
We are currently very strict about breakpoint location matching. This
diff allows some fuzz in the column, but not in the line.

Changelog: [Internal] Setting Hermes breakpoints no longer requires exact column match

Reviewed By: avp

Differential Revision: D21343198

fbshipit-source-id: a59786a9d63f9fe1ed576835ed660ba3343affe1

tmikov committed:
EASY: Allow Unicode line terminators in strings

Summary:
ES10 allows U+2028 (LINE SEPARATOR), U+2029 (PARAGRAPH SEPARATOR) in
string literals for compatibility with JSON. Remove the existing check.
Add a test.

Github: Closes #235

Reviewed By: avp

Differential Revision: D21386937

fbshipit-source-id: 0cbf2d2ff449c33f6f1f3b42ad28ed3650650b36

kodafb committed:
Add new disassembly output format

Summary:
Add a new output format to the disassembler, and expose it via the `hbcdump` tool.

The new output format resembles that of an "objdump" tool. This is meant to be consumed by tools that expect such a format, not for humans to read.

Reviewed By: tmikov

Differential Revision: D21285731

fbshipit-source-id: afa7e4ca1beb408acdc28041a98eec527471b682

avp committed:
Allow PseudoHandle casting and easier assignment.

Summary:
`PinnedHermesValue` can now be assigned to from a `PseudoHandle`,
so this allows assignment into registers and such without having
to remember to invalidate the `PseudoHandle` afterwards.

Add `vmcast` and `dyn_vmcast` static functions to `PseudoHandle`
which are analogous to `Handle`.
The use of the latter does require invalidating the initial handle,
so it doesn't really work for assignment in the condition of an `if`
statement directly, but is sufficient for most other use cases.

Reviewed By: tmikov

Differential Revision: D19746814

fbshipit-source-id: 84ed028279c0924b630fad37db728dfa79ca9b96

avp committed:
Overload makeHandle<T> for PseudoHandle<HermesValue>.

Summary:
This will allow `runtime->makeHandle<T>(std::move(pseudoHandle))`
by `vmcast`ing the `pseudoHandle` to `T`.

Reviewed By: tmikov

Differential Revision: D19746824

fbshipit-source-id: 176217f34ef52a15325cee577b73750b98d2c4cc

avp committed:
Specialize CallResult<PseudoHandle<>>.

Summary:
Previously, in debug mode, `CallResult<PseudoHandle<>>` was using
the default `llvm::Optional` constructor, which doesn't work because
`PseudoHandle` doesn't provide the ability to copy.

Specialize `CallResult<PseudoHandle<>>` into a new `PseudoHandleHV`
`CallResultSpecialize` and use the move constructor to allow simpler
use of `CallResult<PseudoHandle<>>`.

Provide a `toCallResultHermesValue` for simpler migration of functions
which must return `CallResult<HermesValue>`. This allows for partially
moving some APIs to `CallResult<PseudoHandle<>>` without having to,
say, change the signature of `NativeFunctionPtr`.

Reviewed By: mhorowitz

Differential Revision: D19746799

fbshipit-source-id: af8b0d6881930635b6e77fcd16f7c00dae519fac

javache committed:
Add virtual destructor to JSError

Summary:
We consume Hermes through multiple .so's, which means we have multiple (weak) typeinfo definitions of facebook::jsi::JSError. Previously we were using gnustl, which would strcmp typeinfo to decide whether a certain exception handler applies, which meant this didn't cause any major issues. However since this is deprecated, we recently switched to libc++, which does not have this by behaviour (or it does, but behind a flag I'm not sure how to enable). This causes any JS exceptions to fall through from our exception handlers and fatal the app.

This problem is actually documented in the common Android NDK problems page: https://android.googlesource.com/platform/ndk/+/master/docs/user/common_problems.md#rtti_exceptions-not-working-across-library-boundaries

The suggested solution is to ensure that any exception types have a key function defined (a non-pure, out-of-line virtual function). The simplest one to add is a virtual destructor. This makes the object file that holds the implementation of the destructor export a non-weak typeinfo definition which will at load time override the other weak versions.

I'm not sure why we're the first to hit this. RN's JSIExecutor doesn't explicitly reference JSError which probably helps (https://github.com/facebook/react-native/blob/master/ReactCommon/jsiexecutor/jsireact/JSIExecutor.cpp#L256-L258) and they also don't use unguarded callbacks like we do.

Changelog: [Internal]

Reviewed By: mhorowitz

Differential Revision: D21426524

fbshipit-source-id: 474284ada1ca2810045dc4402c420879447f9308

avp committed:
Use AllowDiv when doing REPL paren matching.

Summary:
If we use `AllowRegExp` as the grammar context, then paren matching
breaks if we use `/` as division inside parens.

Always use `AllowDiv` instead. Note that this is not perfect,
because it doesn't work if there are unbalanced parens inside
an actual RegExp.

Reviewed By: tmikov

Differential Revision: D21434307

fbshipit-source-id: c3258bbe73d815349985f97629b54d9dc14e007f

avp committed:
Use index as value in StackTracesTreeNode maps.

Summary:
`StackTracesTreeNode` stores various `llvm::DenseMap` which allow it to point
to its children `StackTracesTreeNode *`s.

The need to iterate these children resulted in the creation of
a `ChildrenIterator` which iterated in an unspecified order.

Remove the nondeterminism incurred by iterating `llvm::DenseMap` by
storing all child nodes in an `std::vector` and using indexes as
values in the `llvm::DenseMap`. This also simplifies iteration as we
can iterate `.getChildren()` directly.

The other option was to use `std::map`, but it doesn't provide the same
`try_emplace` API as `llvm::DenseMap` and this is a simpler fix with a
cleaner iteration method.

Reviewed By: dulinriley

Differential Revision: D21505460

fbshipit-source-id: be1ca655eae43f070f9c1d00a875a7fe4ef6182d

yangshun committed:
Fix website Playground loading by removing unnecessary postprocessing (#236)

Summary:
Fixes https://github.com/facebook/hermes/issues/217.

Visiting the URL https://hermesengine.dev/playground/ directly errored out because the asset paths within the file were incorrectly defined as relative paths. Docusaurus would always use absolute paths based on the `baseUrl` defined in `docusaurus.config.js`.

Upon investigation, I found out there was a preprocessing step which was causing the paths to be modified after `docusaurus build`. I looked at the commit history and saw that they were added in the commit which added the Docusaurus 1 website. Not sure why they were needed then but I'm quite sure we don't need them now.
Pull Request resolved: https://github.com/facebook/hermes/pull/236

Test Plan:
Tested going to /playground locally and it seems fine. The playground doesn't load but I think that's expected since I didn't do the Hermes build steps. cc rickhanlonii

Built the site and loaded /playground using a local static file web server.

<img width="1552" alt="Screen Shot 2020-05-05 at 12 59 23 PM" src="https://user-images.githubusercontent.com/1315101/81038205-c3a98100-8ed7-11ea-8b5a-cbb8eef4b83f.png">

Reviewed By: rickhanlonii

Differential Revision: D21398073

Pulled By: yangshun

fbshipit-source-id: eaf95b7aaac4f8cba5d21818d53a54b184d53016

Commit:
Fix fbcode InitLLVM memory leak

Summary:
`llvm::InitLLVM` already does the calls to:
* `PrintStackTraceOnErrorSignal`
* `PrettyStackTraceProgram`
* `llvm_shutdown`

for us, so there's no need to call them all again.
Sometimes when ASAN/LSAN is on, `InitLLVM` can be called
by the ASAN infra, so guard against that.

Reviewed By: tmikov

Differential Revision: D21207147

fbshipit-source-id: deab21e4e431c875521021fcfa18d90af077b05a

avp committed:
Check recursion depth in member select chains.

Summary:
Some parser functions create nested nodes in a loop, and bypass the
`CHECK_RECURSION` macro. Fix the member select chain by manually
incrementing recursion depth and checking it.

Reviewed By: kodafb

Differential Revision: D21527389

fbshipit-source-id: c86832c738dc575f10996878f3c53967c6485d86

Huxpro committed:
Fault-tolerant Date Parsing

Summary:
Making `Date.parse` accepts garbages and arbitrary spaces.
The tolerance level is something higher than SM but lower than JSC/V8,
which supports duplicated month or even dates (V8):

```
λ eshost -e "new Date('Tue May Jan Apr 05 2020 00:00:00')"
#### Hermes
Invalid Date

#### JavaScriptCore
Sun Apr 05 2020 00:00:00 GMT-0700 (PDT)

#### SpiderMonkey
Invalid Date

#### V8
Sun Apr 05 2020 00:00:00 GMT-0700 (Pacific Daylight Time)
```

Updated test cases in `date-constructor.js` demos our tolerance level.

Reviewed By: avp

Differential Revision: D21434683

fbshipit-source-id: 15691ce3be5261dda328af60c929fd194629462d

Commit:
Improve sanityCheck error message for buffer too small

Summary:
There were two duplicate error messages in `sanityCheck`: "Buffer too small".
In the interest of having unique error messages for unique errors, which makes
triaging the cause easier, make them unique.

Also, since they're saying the size is too small, might as well include the actual
and expected sizes.

Reviewed By: JoshuaGross

Differential Revision: D21553722

fbshipit-source-id: 6bf183b81b6492b4079f3d2242d4192ce41d67f7

Commit:
Make GCPointer<T>::get use vmcast

Summary:
There are some extra debug checks that could be done whenever
a pointer is accessed from a `GCPointer<T>`, to make sure it is
actually a `T *`.

This is particularly helpful if the mutator accesses memory that the GC
has freed, since the CellKind will likely be different.

Reviewed By: tmikov

Differential Revision: D21408268

fbshipit-source-id: cdc038b743bb1a7f14ae04ba7777bec7e0982090

Commit:
Add ASAN poisoning to Hades

Summary:
To get some additional checking of out-of-bounds memory in Hades,
poison memory ranges that are represented by free memory.

The only parts left unpoisoned are the cell header so that types can
be checked and the free list can be iterated.

Alternatively, the poison could be cleared by the GC whenever it reads a cell
that might be free, but that would be much more computationally expensive.

Reviewed By: davedets

Differential Revision: D20700574

fbshipit-source-id: 1c28eb8dcacd62f5467dee9c28a50af3bf4e3076

Commit:
Add TSAN function declarations to hermes/Support/Compiler.h

Summary:
TSAN has some pre-defined functions you can call to let TSAN know
extra information. LLVM had some of these defined, but was missing others.
Add some useful ones like for ignoring races, and for setting the thread name.
Add the functions in hermes instead of LLVM to avoid maintaining patches.

These function headers are specified in a couple places, but I found the most
reliable source of information was Abseil:
https://github.com/abseil/abseil-cpp/blob/master/absl/base/dynamic_annotations.h

Reviewed By: tmikov

Differential Revision: D21304897

fbshipit-source-id: 5b9187bf246bf25bc4969c31956e6298adc85f00

Commit:
Add set_thread_name to OSCompat

Summary:
Setting a thread name changes based on the platform, and should go into
OSCompat. Add a Linux, Android, MacOS, and Windows implementation.

Reviewed By: tmikov

Differential Revision: D21304910

fbshipit-source-id: ce8c3a5d5b8f67708d4acda070e63f68264e543e

Commit:
Implement sequential collection for HadesGC

Summary:
Implement allocation into young gen, evacuation of young gen, and collection of old gen.
All of these are implemented sequentially right now, to get the basic structure working and tests passing.
Currently only marking and sweeping are done in the old gen, no compaction is done at all.

Concurrency will be added in later stages. It is expected that there will probably be some major changes to data structures and algorithms used during those changes.
For those reasons I have left some performance-related work as "TODO".

Currently this GC is not yet suitable to be activated, it is 2x slower on v8-splay (probably mostly from the naive free-list implementation), and is missing some fringe functionality (snapshots, tripwire, clearing DPM, etc.).

Reviewed By: davedets

Differential Revision: D20241740

fbshipit-source-id: 685eaa1e7bb9ad255e1b2a61efed270ddd29e3d3

Commit:
Allow correct count of parentheses in regex

Summary:
Parenthesis inside a regular expression should not be taken into account when balancing parenthesis in a line. For example, `/foo\(/` is balanced.

The problem is that it is hard to tell if `/` is used for division or a regular expression.
* Division
  * `1 / foo / 1`
* Regular Expression
  * `/foo/`

https://github.com/mikesamuel/es5-lexer/blob/master/src/guess_is_regexp.js uses a heuristic to decide this by looking at the previous token and guessing if the current `/` is used for division or the beginning of a regular expression.

Reviewed By: avp

Differential Revision: D21528909

fbshipit-source-id: 6cac28191d94f16194d4548b575ea93a3cc1af5c

Commit:
Migrate *-fbcode to *Fbcode

Reviewed By: mzlee

Differential Revision: D21183270

fbshipit-source-id: 54ad3a1704a97d007aea1743755c541a0c3683f1

avp committed:
Avoid memsetting new slabs in BacktrackingBumpPtrAllocator.

Summary:
`allocateNewSlab` was using `llvm::make_unique`, which calls the
constructor on the template parameter, and forces a `memset` on the
whole new slab.

We don't care if the new slab is zero-initialized because we're doing
bump pointer allocation and we will overwrite the allocated memory with
something useful any way.

Use `new Slab` without parentheses and avoid `make_unique` to prevent
`memset` from running.

Reviewed By: tmikov

Differential Revision: D21575586

fbshipit-source-id: 513587f75e092007d4d966477f2f23816b75bf84

Huxpro committed:
Fix scrolling

Summary: <https://github.com/facebook/hermes/issues/245>

Reviewed By: tmikov

Differential Revision: D21578811

fbshipit-source-id: 93ac0b3e65b776d4afaa6a643c3b26749c294313

avp committed:
Fix handle leak in JSMapImpl::forEach.

Summary:
`forEach` wasn't flushing handles through its loop, causing unbounded
handle creation in certain cases.

Fix and add a test.

Reviewed By: dulinriley

Differential Revision: D21576611

fbshipit-source-id: 63f02083104314a50a3eabfd0b86efdf4f462e81

willholen committed:
Add error messages for non-parsing source map errors

Summary:
Currently, if a field is missing from the source map, there's no error message:

    $ hermes ~/current.js -source-map ~/current.js.map -emit-binary -out ~/current.hbc
    $

This diff adds error messages in these cases:

    $ hermes ~/current.js -source-map ~/current.js.map -emit-binary -out ~/current.hbc
    /Users/willholen/current.js.map:1:1: error: 'sources' key missing from source map
    {
    ^
    $

Since we don't currently track source locations for JSON objects, these warnings
are emitted on the first byte of the file.

Reviewed By: avp

Differential Revision: D21580417

fbshipit-source-id: 4a1aa2cca489b9e14eaf198371f98c152d603e1c

mhorowitz committed:
Add support for additional internal property slots in DecoratedObject

Summary:
This corresponds to ecma262 OrdinaryObjectCreate's
additionalInternalSlotsList.

Reviewed By: tmikov

Differential Revision: D21420113

fbshipit-source-id: 4d718f1f5d05dfda7e588d373b8885d1541ecd5f

mhorowitz committed:
Improve comments on createListFromArrayLike

Reviewed By: dulinriley

Differential Revision: D21420116

fbshipit-source-id: 2e433753ac2545b0fae0f7f44a89bd6c5d815a43

mhorowitz committed:
Replace JSObject* in arg list with NativeFunction*

Reviewed By: tmikov

Differential Revision: D21420110

fbshipit-source-id: 5fd17ee4ebf58dafbe3b335ca0dc2c39dd03f2e6

mhorowitz committed:
Add toNumeric_RJS

Summary: This was added in newer versions of 262.

Reviewed By: avp

Differential Revision: D21420112

fbshipit-source-id: ffd504ab848c517ef3c1a8e5f87d6b1ec8fd537c

avp committed:
Use emplace_back in allocateNewSlab.

Summary:
emplace_back emits slightly less code than push_back in some cases,
so use it.

Continue to use `new Slab` to avoid zeroing out memory.

Reviewed By: dulinriley

Differential Revision: D21581061

fbshipit-source-id: 94af27d0ee71d4c4ea36268180c362a798cae741

Commit:
Fix unit tests for Hades, and add it to CI

Summary:
Some unit tests don't work with Hades enabled, so apply fixes.
Mostly these are just ifdef'ing out the test because it uses some debug
counters that aren't available in Hades, or otherwise makes assumptions about
how/when functions are called.

Reviewed By: davedets

Differential Revision: D21284417

fbshipit-source-id: 21ac4331a363ad1170281d0e4ef91b8177f3f46e

Commit:
Fix WeakMap bug where a map is only found in the final drain

Summary:
A crash was discovered in Hermes that looks like this:
```
function main() {
  var v1 = {};
  var wm1 = new WeakMap();
  var v2 = {};
  wm1.set(v2, v1);
  var wm2 = new WeakMap();
  v1.a = wm2;
  var v3 = [0];
  for (var i = 0; i < 57; i++) {
    wm1.set(v3, 0);
    v3 = [0];
    v3[3262] = 0;
  }
}
main();
```

The crash happens because `wm2`'s valueStorage_ member wasn't getting marked.

`wm1` is reachable. It has a key/value pair where the key for `v2` has been cleared,
and the value field points to `v1`. `wm2` is only reachable via `v1` when bytecode
optimization is turned on, because the registers for `v1` and `v2` are re-used for
other values. A YG GC nulls out the key in `wm1`, since `v2` is garbage early on.

Once a full GC occurs and it finishes marking, it only discovers `wm1`.
`markFromReachableWeakMapKeys` does not mark the value fields of pairs whose key is null,
so `v1` (and thus `wm2`) is not discovered.
But `clearEntriesWithUnreachableKeys` didn't have the same behavior; it didn't clear this entry.
So `v1` and `wm2` were discovered and marked during the end of weak map marking.

The crash occurs because `wm2` was now marked as reachable, but had the marking of
its internal array delayed until the next `drainMarkStack`, so the internal array was not marked
reachable and was swept away. The next access, or next GC that followed the pointer again,
could read garbage memory.

The solution is easy: treat null pointers as unreachable in `clearEntriesWithUnreachableKeys`.

To find this problem faster in the future, I added an assert to make sure no new WeakMaps are
found as a result of the final `drainMarkStack`.

Reviewed By: davedets

Differential Revision: D21537994

fbshipit-source-id: ddb38f048c36578eb7854216c7406e982909a20e

avp committed:
Avoid IRGen during -dump-ast when -commonjs is provided.

Summary:
A previous diff stopped compilation early but neglected `-commonjs` mode.
This fixes that.

Reviewed By: dulinriley

Differential Revision: D20599904

fbshipit-source-id: 34060ab2c40f1409f94acc958d8ef4cea49fbf10

avp committed:
Move async function checks to IRGen.

Summary:
We don't support generating code for async functions, but they
are semantically valid and we can parse them,
so we shouldn't emit errors during `-dump-ast`.

Reviewed By: tmikov

Differential Revision: D20770236

fbshipit-source-id: ff8dca29c33c14808a2927b02c713713a09bd655

willholen committed:
Treat an empty source map string as missing

Summary:
Currently, an empty string will be treated as a source map, while nullptr will
treat it as missing. For easier JS interop, treat these the same.

Reviewed By: ArchDev

Differential Revision: D21621982

fbshipit-source-id: c408b3889ce368ce4bad5b4529deca8c1eeb9c40

avp committed:
Add ESTREE_NODE_7_ARGS and ESTREE_NODE_8_ARGS.

Summary: These are needed for Flow types.

Reviewed By: dulinriley

Differential Revision: D20562407

fbshipit-source-id: 2175e7f4014b1ace6be9d2d0ef6b00c2167418a7

avp committed:
Remove some restrictions on JSLexer lookahead.

Summary:
For `import()` and Flow type parsing we need more lookahead in our
JSLexer, so rename `lookaheadAfterAsync` to `lookahead1` and allow
the current token to be an identifier or a reserved word.

Reviewed By: tmikov

Differential Revision: D20562405

fbshipit-source-id: 82d6adce24dfbb92b23d62e01b4196c2b35b5617

avp committed:
Parse JSX.

Summary:
Add a `-jsx` mode to Hermes which can parse JSX into ESTree.

Unsupported:
- Raw tags as attributes, as support across parsers is inconsistent right
  now, and the lexer needs to be taught to understand `>>` as two
  separate `>` and `>` tokens in JSX mode.
- Converting XHTML entities into actual symbols (`&lt;` into `<`, etc)
  will be added in a future diff.

The tricky part here is the two lexer changes:
- `advanceInJSXChild` allows consuming `JSXText`, which is introduced
  as a new token kind. All characters except `{`,`}`,`<`,`>` are valid
  in JSXText so we add a separate function for it. We introduce a depth
  counter for JSX elements so we know when to continue scanning for
  JSX text after parsing a closing tag.
- `advance` is given a new `GrammarContext` which changes the way
  that identifiers are scanned. `JSXIdentifier` can contain but not begin
  with `-`, so we templatize a few functions to allow this while not
  adding more branches to the hot path for JS identifier scanning.

Reviewed By: tmikov

Differential Revision: D20173526

fbshipit-source-id: f54ab27e9ce929a7eacc7417e91ef3bba2e5ec10

avp committed:
Use early returns in ASTBuilder.

Summary: Avoid too many nested if/else clauses, which VC++ has issues with.

Reviewed By: dulinriley

Differential Revision: D21002495

fbshipit-source-id: 87e1f0c42b7965f6ccd649e86f042f9c2b0c002e

avp committed:
ESTree for JSX.

Summary: Based on https://github.com/facebook/jsx/blob/master/AST.md

Reviewed By: tmikov

Differential Revision: D20173527

fbshipit-source-id: 84e1255fdf292cb861eed12f8c9395546fa52fa8

avp committed:
Parse import() calls.

Summary:
The `import` resword can only be used as part of a call, so check that
before calling `parseCallExpression`. We also need to lookahead before
calling `parseImportDeclaration` because `import` is no longer an indication
that we are definitely parsing an import declaration.

Reviewed By: tmikov

Differential Revision: D20770237

fbshipit-source-id: 37cb91be2d27b9e3b08b436b98e15656191691f8

Commit:
Fix invalid unicode test

Summary:
The orignal test was not maching the expected lines correctly.

This is the original `unicode-invalid.js`
```
૬͓͋ = false;
// Starts with a digit with combining mark
// CHECK: unrecognized Unicode character \u34b

﹏ = false;
// Starts with connector puncutation
// CHECK: unrecognized Unicode character \u353
```

And the full output of hermes
```
/data/users/ellishoag/fbsource/xplat/hermes/test/Parser/unicode-invalid.js:13:2: error: unrecognized Unicode character \u35b
 ͛xyz = false;
/data/users/ellishoag/fbsource/xplat/hermes/test/Parser/unicode-invalid.js:17:1: error: unrecognized Unicode character \u194a
᥊abc = false;
/data/users/ellishoag/fbsource/xplat/hermes/test/Parser/unicode-invalid.js:21:1: error: unrecognized Unicode character \uaec
૬͓͋ = false;
/data/users/ellishoag/fbsource/xplat/hermes/test/Parser/unicode-invalid.js:21:4: error: unrecognized Unicode character \u34b
૬͓͋ = false;
/data/users/ellishoag/fbsource/xplat/hermes/test/Parser/unicode-invalid.js:21:6: error: unrecognized Unicode character \u353
૬͓͋ = false;
/data/users/ellishoag/fbsource/xplat/hermes/test/Parser/unicode-invalid.js:21:9: error: invalid expression
૬͓͋ = false;
```

The last line in the file is not actually reported as an error.

Reviewed By: tmikov

Differential Revision: D21602383

fbshipit-source-id: 8a47b035a625e053d4f442b38fe616bf460a1742
